File: camerax_core/lib/src/instance_manager.dart
abstract class InstanceManager {
  int tryAddInstance<T>(T instance);
  void removeInstance<T>(T instance);
  T getInstance<T>(int instanceId);

  static final InstanceManager instance = _InstanceManager();
}

class _InstanceManager implements InstanceManager {
  final instances = <int, dynamic>{};
  final instanceIds = <dynamic, int>{};

  var instanceId = 0;

  @override
  int tryAddInstance<T>(T instance) {
    var instanceId = instanceIds[instance];
    if (instanceId == null) {
      instanceId = this.instanceId++;
      instanceIds[instance] = instanceId;
      instances[instanceId] = instance;
    }
    return instanceId;
  }

  @override
  void removeInstance<T>(T instance) {
    final instanceId = instanceIds[instance];
    instanceIds.remove(instance);
    instances.remove(instanceId);
  }

  @override
  T getInstance<T>(int instanceId) {
    return instances[instanceId];
  }
}

File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/InstanceManager.kt
package dev.hebei.camerax_android

import java.lang.ref.WeakReference

typealias FinalizationListener = (identifier: Long) -> Unit

@Suppress("UNCHECKED_CAST")
object InstanceManager {
    private val instances = mutableMapOf<Long, Any>()

    private fun add(id: Long, instance: Any) {
        if (instances.containsKey(id)) {
            throw IllegalArgumentException(String.format("Identifier has already been added: %d", id))
        }
        instances[id] = instance
    }

    fun <T> remove(identifier: Long): T? {
        return instances.remove(identifier) as T?
    }

    fun <T> getInstance(identifier: Long): T? {
        return instances[identifier] as T?
    }

    fun clear() {
        instances.clear()
    }
}